# GRADED CHALLENGE

- Nama : Nicholas Haryadi
- Batch : HCK-005

Pada notebook ini akan dilakukan analisis terhadap ...

## IMPORT LIBRARY

In [18]:
import pandas as pd


## DATA LOADING AND OVERVIEW


In [ ]:
df = pd.read_csv(r'C:\Users\HP\Desktop\PandasNBA\dataset\nbaallelo.csv.csv')

In [24]:
df.head(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   

   is_playoffs team_id  fran_id  ...  win_equiv  opp_id  opp_fran  opp_pts  \
0            0     TRH  Huskies  ...   40.29483     NYK    Knicks       68   
1            0     NYK   Knicks  ...   41.70517     TRH   Huskies       66   

  opp_elo_i  opp_elo_n  game_location  game_result  forecast notes  
0    1300.0  1306.7233              H            L  0.640065   NaN  
1    1300.0  1293.2767              A            W  0.359935   NaN  

[2 rows x 23 columns]

In [25]:
df.tail(2)

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
126312         102            1     CLE  Cavaliers  ...  59.290245     GSW   
126313         103            1     GSW   Warriors  ...  68.519516     CLE   

         opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
126312   Warriors      105  1813.6349  1822.2881              H            L   
126313  Cavaliers       97  1700.7391  1692.0859              A            W   

        forecast notes  
126312   0.48145   NaN  
126313   0.51855   NaN  

[2 rows x 23 columns]

In [27]:
df.sample(5)

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
53775       26888  198511050UTA   NBA        1     1986   11/5/1985   
66550       33276  199105030IND   NBA        0     1991    5/3/1991   
26970       13486  197211180CHI   NBA        0     1973  11/18/1972   
114599      57300  201012300DAL   NBA        0     2011  12/30/2010   
113306      56654  201004030NJN   NBA        0     2010    4/3/2010   

        seasongame  is_playoffs team_id    fran_id  ...  win_equiv  opp_id  \
53775            6            0     GSW   Warriors  ...  26.751680     UTA   
66550           86            1     IND     Pacers  ...  44.427731     BOS   
26970           16            0     CHI      Bulls  ...  52.619053     BAL   
114599          31            0     DAL  Mavericks  ...  54.905243     SAS   
113306          76            0     NJN       Nets  ...  21.996107     NOH   

        opp_fran  opp_pts  opp_elo_i  opp_elo_n  game_location  game_result  \
53775       Jazz      118  1505.8397  1511.1777              A            L   
66550    Celtics      113  1545.2261  1540.9459              H            W   
26970    Wizards       79  1458.7234  1453.4279              H            W   
114599     Spurs       99  1678.7986  1688.7014              H            L   
113306  Pelicans       87  1439.7408  1414.3027              H            W   

        forecast notes  
53775   0.201193   NaN  
66550   0.596426   NaN  
26970   0.778729   NaN  
114599  0.603089   NaN  
113306  0.416953   NaN  

[5 rows x 23 columns]

In [28]:
pd.set_option("display.max.columns", None)

In [29]:
df.head(5)

gameorder       game_id lg_id  _iscopy  year_id  date_game  seasongame  \
0          1  194611010TRH   NBA        0     1947  11/1/1946           1   
1          1  194611010TRH   NBA        1     1947  11/1/1946           1   
2          2  194611020CHS   NBA        0     1947  11/2/1946           1   
3          2  194611020CHS   NBA        1     1947  11/2/1946           2   
4          3  194611020DTF   NBA        0     1947  11/2/1946           1   

   is_playoffs team_id  fran_id  pts      elo_i      elo_n  win_equiv opp_id  \
0            0     TRH  Huskies   66  1300.0000  1293.2767  40.294830    NYK   
1            0     NYK   Knicks   68  1300.0000  1306.7233  41.705170    TRH   
2            0     CHS    Stags   63  1300.0000  1309.6521  42.012257    NYK   
3            0     NYK   Knicks   47  1306.7233  1297.0712  40.692783    CHS   
4            0     DTF  Falcons   33  1300.0000  1279.6189  38.864048    WSC   

   opp_fran  opp_pts  opp_elo_i  opp_elo_n game_location game_result  \
0    Knicks       68  1300.0000  1306.7233             H           L   
1   Huskies       66  1300.0000  1293.2767             A           W   
2    Knicks       47  1306.7233  1297.0712             H           W   
3     Stags       63  1300.0000  1309.6521             A           L   
4  Capitols       50  1300.0000  1320.3811             H           L   

   forecast notes  
0  0.640065   NaN  
1  0.359935   NaN  
2  0.631101   NaN  
3  0.368899   NaN  
4  0.640065   NaN

In [30]:
df.shape

(126314, 23)

In [32]:
df.info

<bound method DataFrame.info of         gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
1               1  194611010TRH   NBA        1     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
3               2  194611020CHS   NBA        1     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts      elo_i      elo_n  \
0                1            0     TRH    Hus

In [33]:
df.describe().T

count          mean           std          min           25%  \
gameorder    126314.0  31579.000000  18231.927643     1.000000  15790.000000   
_iscopy      126314.0      0.500000      0.500002     0.000000      0.000000   
year_id      126314.0   1988.200374     17.582309  1947.000000   1975.000000   
seasongame   126314.0     43.533733     25.375178     1.000000     22.000000   
is_playoffs  126314.0      0.063857      0.244499     0.000000      0.000000   
pts          126314.0    102.729982     14.814845     0.000000     93.000000   
elo_i        126314.0   1495.236055    112.139945  1091.644500   1417.237975   
elo_n        126314.0   1495.236055    112.461687  1085.774400   1416.994900   
win_equiv    126314.0     41.707889     10.627332    10.152501     34.103035   
opp_pts      126314.0    102.729982     14.814845     0.000000     93.000000   
opp_elo_i    126314.0   1495.236055    112.139945  1091.644500   1417.237975   
opp_elo_n    126314.0   1495.236055    112.461687  1085.774400   1416.994900   
forecast     126314.0      0.500000      0.215252     0.020447      0.327989   

                      50%           75%           max  
gameorder    31579.000000  47368.000000  63157.000000  
_iscopy          0.500000      1.000000      1.000000  
year_id       1990.000000   2003.000000   2015.000000  
seasongame      43.000000     65.000000    108.000000  
is_playoffs      0.000000      0.000000      1.000000  
pts            103.000000    112.000000    186.000000  
elo_i         1500.945550   1576.060000   1853.104500  
elo_n         1500.954400   1576.291625   1853.104500  
win_equiv       42.113357     49.635328     71.112038  
opp_pts        103.000000    112.000000    186.000000  
opp_elo_i     1500.945550   1576.060000   1853.104500  
opp_elo_n     1500.954400   1576.291625   1853.104500  
forecast         0.500000      0.672011      0.979553

In [46]:
df[['team_id', 'fran_id', 'pts']]

team_id    fran_id  pts
0          TRH    Huskies   66
1          NYK     Knicks   68
2          CHS      Stags   63
3          NYK     Knicks   47
4          DTF    Falcons   33
...        ...        ...  ...
126309     CLE  Cavaliers   82
126310     GSW   Warriors  104
126311     CLE  Cavaliers   91
126312     CLE  Cavaliers   97
126313     GSW   Warriors  105

[126314 rows x 3 columns]

In [47]:
df['team_id'].unique() #JUMLAH UNIQUE

array(['TRH', 'NYK', 'CHS', 'DTF', 'WSC', 'BOS', 'PRO', 'PIT', 'STB',
       'CLR', 'PHW', 'BLB', 'INJ', 'FTW', 'MNL', 'ROC', 'TRI', 'DNN',
       'INO', 'SHE', 'WAT', 'AND', 'SYR', 'MLH', 'STL', 'DET', 'CIN',
       'LAL', 'CHP', 'CHZ', 'SFW', 'BAL', 'PHI', 'CHI', 'SEA', 'OAK',
       'ANA', 'SDR', 'KEN', 'INA', 'DNR', 'DLC', 'HSM', 'NOB', 'MNM',
       'NJA', 'PTP', 'ATL', 'MIL', 'PHO', 'NYA', 'MMF', 'MNP', 'LAS',
       'CAR', 'WSA', 'BUF', 'CLE', 'UTS', 'FLO', 'PTC', 'POR', 'VIR',
       'TEX', 'MMP', 'HOU', 'GSW', 'KCO', 'MMT', 'SDA', 'CAP', 'SAA',
       'NOJ', 'DNA', 'SSL', 'MMS', 'WSB', 'KCK', 'SDS', 'IND', 'NYN',
       'DEN', 'SAS', 'NJN', 'SDC', 'UTA', 'DAL', 'LAC', 'SAC', 'CHH',
       'MIA', 'MIN', 'ORL', 'VAN', 'TOR', 'WAS', 'MEM', 'NOH', 'CHA',
       'NOK', 'OKC', 'BRK', 'NOP', 'CHO'], dtype=object)

In [49]:
df['team_id'].nunique() #JUMLAH TABEL YG UNIQUE

104

In [52]:
df['team_id'].value_counts().head() #head buat paling 5 teratas

BOS    5997
NYK    5769
LAL    5078
DET    4985
PHI    4533
Name: team_id, dtype: int64

In [53]:
df['fran_id'].value_counts().head()

Lakers      6024
Celtics     5997
Knicks      5769
Warriors    5657
Pistons     5650
Name: fran_id, dtype: int64

Terdapat ketidak konsistensian pada data dimana kolom fran_id menunjukkan

In [64]:
# .loc & .iloc
# loc = label terlampir ( seng diatas kaya year_id,dll )
# iloc = index penomoran
df.loc[df['fran_id'] == 'Lakers', 'team_id'].value_counts()

LAL    5078
MNL     946
Name: team_id, dtype: int64

In [66]:
5078 + 946 

6024

Diketahui bahwa Lakers memiliki 2 kode yaitu LAL dan MNL sehingga setelah dari 2 kode itu dijumlahkan maka Lakers memiliki jumlah point 6024.

In [68]:
df.loc[df['fran_id'] == 'Lakers', 'pts'].mean()

105.81739707835325

In [70]:
df.iloc[:, 10]

0          66
1          68
2          63
3          47
4          33
         ... 
126309     82
126310    104
126311     91
126312     97
126313    105
Name: pts, Length: 126314, dtype: int64

## QUERY

In [73]:
df[df['fran_id'] == 'Lakers'][['fran_id', 'team_id']]

fran_id team_id
1136    Lakers     MNL
1152    Lakers     MNL
1159    Lakers     MNL
1170    Lakers     MNL
1183    Lakers     MNL
...        ...     ...
126016  Lakers     LAL
126052  Lakers     LAL
126086  Lakers     LAL
126115  Lakers     LAL
126137  Lakers     LAL

[6024 rows x 2 columns]

In [80]:
# fran_id = Warriors
# year_id = diatas 2010
# points > 100
# game location away
df [(df['fran_id'] == 'Warriors') &
   (df['year_id']> 2010) &
   (df['pts']>100) &
   (df['game_location'] == 'A')]

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
113853      56927  201011080TOR   NBA        1     2011   11/8/2010   
113874      56938  201011100NYK   NBA        1     2011  11/10/2010   
114076      57039  201011240HOU   NBA        1     2011  11/24/2010   
114106      57054  201011260MEM   NBA        1     2011  11/26/2010   
114127      57064  201011270MIN   NBA        1     2011  11/27/2010   
...           ...           ...   ...      ...      ...         ...   
126280      63141  201505150MEM   NBA        1     2015   5/15/2015   
126292      63147  201505230HOU   NBA        1     2015   5/23/2015   
126296      63149  201505250HOU   NBA        1     2015   5/25/2015   
126308      63155  201506110CLE   NBA        1     2015   6/11/2015   
126313      63157  201506170CLE   NBA        1     2015   6/16/2015   

        seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
113853           7            0     GSW  Warriors  109  1449.8333  1461.3705   
113874           8            0     GSW  Warriors  122  1461.3705  1470.0310   
114076          15            0     GSW  Warriors  101  1426.7416  1420.6514   
114106          16            0     GSW  Warriors  111  1420.6514  1416.7081   
114127          17            0     GSW  Warriors  104  1416.7081  1426.9871   
...            ...          ...     ...       ...  ...        ...        ...   
126280          92            1     GSW  Warriors  108  1777.6378  1787.3469   
126292          95            1     GSW  Warriors  115  1791.2300  1813.4543   
126296          96            1     GSW  Warriors  115  1813.4543  1797.8933   
126308         101            1     GSW  Warriors  103  1790.9591  1809.9791   
126313         103            1     GSW  Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
113853  36.593380    TOR       Raptors      102  1444.9259  1433.3887   
113874  37.493332    NYK        Knicks      117  1421.5227  1412.8622   
114076  32.433086    HOU       Rockets      111  1454.8855  1460.9757   
114106  32.038364    MEM     Grizzlies      116  1458.7405  1462.6837   
114127  33.070763    MIN  Timberwolves       94  1314.1195  1303.8406   
...           ...    ...           ...      ...        ...        ...   
126280  66.383080    MEM     Grizzlies       95  1618.0546  1608.3455   
126292  68.002609    HOU       Rockets       80  1665.6567  1643.4324   
126296  67.053963    HOU       Rockets      128  1643.4324  1658.9934   
126308  67.794983    CLE     Cavaliers       82  1723.4149  1704.3949   
126313  68.519516    CLE     Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
113853             A           W  0.366468   NaN  
113874             A           W  0.414289   NaN  
114076             A           L  0.323517   NaN  
114106             A           L  0.311116   NaN  
114127             A           W  0.503725   NaN  
...              ...         ...       ...   ...  
126280             A           W  0.584916   NaN  
126292             A           W  0.536737   NaN  
126296             A           L  0.599427   NaN  
126308             A           W  0.453428   NaN  
126313             A           W  0.518550   NaN  

[116 rows x 23 columns]

In [81]:
# fran_id = Warriors
# year_id = diatas 2010
# points > 100
# game location away
df.query('fran_id == "Warriors" \
         & year_id > 2010 \
         & pts > 100 \
         & game_location == "A"')

gameorder       game_id lg_id  _iscopy  year_id   date_game  \
113853      56927  201011080TOR   NBA        1     2011   11/8/2010   
113874      56938  201011100NYK   NBA        1     2011  11/10/2010   
114076      57039  201011240HOU   NBA        1     2011  11/24/2010   
114106      57054  201011260MEM   NBA        1     2011  11/26/2010   
114127      57064  201011270MIN   NBA        1     2011  11/27/2010   
...           ...           ...   ...      ...      ...         ...   
126280      63141  201505150MEM   NBA        1     2015   5/15/2015   
126292      63147  201505230HOU   NBA        1     2015   5/23/2015   
126296      63149  201505250HOU   NBA        1     2015   5/25/2015   
126308      63155  201506110CLE   NBA        1     2015   6/11/2015   
126313      63157  201506170CLE   NBA        1     2015   6/16/2015   

        seasongame  is_playoffs team_id   fran_id  pts      elo_i      elo_n  \
113853           7            0     GSW  Warriors  109  1449.8333  1461.3705   
113874           8            0     GSW  Warriors  122  1461.3705  1470.0310   
114076          15            0     GSW  Warriors  101  1426.7416  1420.6514   
114106          16            0     GSW  Warriors  111  1420.6514  1416.7081   
114127          17            0     GSW  Warriors  104  1416.7081  1426.9871   
...            ...          ...     ...       ...  ...        ...        ...   
126280          92            1     GSW  Warriors  108  1777.6378  1787.3469   
126292          95            1     GSW  Warriors  115  1791.2300  1813.4543   
126296          96            1     GSW  Warriors  115  1813.4543  1797.8933   
126308         101            1     GSW  Warriors  103  1790.9591  1809.9791   
126313         103            1     GSW  Warriors  105  1813.6349  1822.2881   

        win_equiv opp_id      opp_fran  opp_pts  opp_elo_i  opp_elo_n  \
113853  36.593380    TOR       Raptors      102  1444.9259  1433.3887   
113874  37.493332    NYK        Knicks      117  1421.5227  1412.8622   
114076  32.433086    HOU       Rockets      111  1454.8855  1460.9757   
114106  32.038364    MEM     Grizzlies      116  1458.7405  1462.6837   
114127  33.070763    MIN  Timberwolves       94  1314.1195  1303.8406   
...           ...    ...           ...      ...        ...        ...   
126280  66.383080    MEM     Grizzlies       95  1618.0546  1608.3455   
126292  68.002609    HOU       Rockets       80  1665.6567  1643.4324   
126296  67.053963    HOU       Rockets      128  1643.4324  1658.9934   
126308  67.794983    CLE     Cavaliers       82  1723.4149  1704.3949   
126313  68.519516    CLE     Cavaliers       97  1700.7391  1692.0859   

       game_location game_result  forecast notes  
113853             A           W  0.366468   NaN  
113874             A           W  0.414289   NaN  
114076             A           L  0.323517   NaN  
114106             A           L  0.311116   NaN  
114127             A           W  0.503725   NaN  
...              ...         ...       ...   ...  
126280             A           W  0.584916   NaN  
126292             A           W  0.536737   NaN  
126296             A           L  0.599427   NaN  
126308             A           W  0.453428   NaN  
126313             A           W  0.518550   NaN  

[116 rows x 23 columns]

## GROUPBY & AGGREGATION


In [86]:
df.groupby('game_location')['pts'].mean()

game_location
A    100.856616
H    104.602395
N    104.315789
Name: pts, dtype: float64

In [95]:
grup = df.groupby('game_location')

In [96]:
grup.get_group('A')['pts'].mean()

100.85661566726853

In [92]:
grup.get_group('H')['pts'].mean()

104.60239475434761

In [93]:
grup.get_group('N')['pts'].mean()

104.3157894736842

In [94]:
df.groupby('game_location').agg({'pts':'mean','year_id':'max'})

pts  year_id
game_location                     
A              100.856616     2015
H              104.602395     2015
N              104.315789     2015

In [97]:
df.groupby(['game_location', 'game_result'])['pts'].mean()

game_location  game_result
A              L               97.221600
               W              106.851424
H              L               97.558763
               W              108.873375
N              L               98.473684
               W              110.157895
Name: pts, dtype: float64

In [99]:
#hitung rata rata point BOSTON vs CHICAGO di tiap tahun dari 2010-2015
df[(df['fran_id'] == 'Celtics') &
    (df['opp_fran'] == 'Bulls') &
    (df['year_id'] >= 2010) &
    (df['year_id'] <= 2015)].groupby('year_id')['pts'].mean()

year_id
2010    100.00
2011     93.50
2012     85.00
2013     90.00
2014     88.00
2015    103.75
Name: pts, dtype: float64

## DATA MANIPULATION
- Rename Column
- Drop Column
- Create new Column

In [102]:
#rename columns
df.rename(columns={'game_location':'loc',
                    'game_result':'result'}, inplace=True)

In [105]:
#drop columns
df.drop(columns=['elo_i','elo_n','notes'], inplace=True)

In [111]:
#drop index (baris)
df.drop(index=[1,3,5,7],axis=0, inplace=True)

KeyError: '[1, 3, 5, 7] not found in axis'

In [114]:
df['selisih'] = df['pts']- df['opp_pts']

In [115]:
df

gameorder       game_id lg_id  _iscopy  year_id  date_game  \
0               1  194611010TRH   NBA        0     1947  11/1/1946   
2               2  194611020CHS   NBA        0     1947  11/2/1946   
4               3  194611020DTF   NBA        0     1947  11/2/1946   
6               4  194611020PRO   NBA        1     1947  11/2/1946   
8               5  194611020STB   NBA        1     1947  11/2/1946   
...           ...           ...   ...      ...      ...        ...   
126309      63155  201506110CLE   NBA        0     2015  6/11/2015   
126310      63156  201506140GSW   NBA        0     2015  6/14/2015   
126311      63156  201506140GSW   NBA        1     2015  6/14/2015   
126312      63157  201506170CLE   NBA        0     2015  6/16/2015   
126313      63157  201506170CLE   NBA        1     2015  6/16/2015   

        seasongame  is_playoffs team_id    fran_id  pts  win_equiv opp_id  \
0                1            0     TRH    Huskies   66  40.294830    NYK   
2                1            0     CHS      Stags   63  42.012257    NYK   
4                1            0     DTF    Falcons   33  38.864048    WSC   
6                1            0     BOS    Celtics   53  40.459381    PRO   
8                1            0     PIT    Ironmen   51  40.507980    STB   
...            ...          ...     ...        ...  ...        ...    ...   
126309         100            1     CLE  Cavaliers   82  60.309792    GSW   
126310         102            1     GSW   Warriors  104  68.013329    CLE   
126311         101            1     CLE  Cavaliers   91  60.010067    GSW   
126312         102            1     CLE  Cavaliers   97  59.290245    GSW   
126313         103            1     GSW   Warriors  105  68.519516    CLE   

            opp_fran  opp_pts  opp_elo_i  opp_elo_n loc result  forecast  \
0             Knicks       68  1300.0000  1306.7233   H      L  0.640065   
2             Knicks       47  1306.7233  1297.0712   H      W  0.631101   
4           Capitols       50  1300.0000  1320.3811   H      L  0.640065   
6       Steamrollers       59  1300.0000  1305.1542   A      L  0.359935   
8            Bombers       56  1300.0000  1304.6908   A      L  0.359935   
...              ...      ...        ...        ...  ..    ...       ...   
126309      Warriors      103  1790.9591  1809.9791   H      L  0.546572   
126310     Cavaliers       91  1704.3949  1700.7391   H      W  0.765565   
126311      Warriors      104  1809.9791  1813.6349   A      L  0.234435   
126312      Warriors      105  1813.6349  1822.2881   H      L  0.481450   
126313     Cavaliers       97  1700.7391  1692.0859   A      W  0.518550   

        selisih  
0            -2  
2            16  
4           -17  
6            -6  
8            -5  
...         ...  
126309      -21  
126310       13  
126311      -13  
126312       -8  
126313        8  

[126310 rows x 21 columns]